tests/pf_gp_exceptions/host/host.c

```diff
@@ -0,0 +1,67 @@
+// Copyright (c) Open Enclave SDK contributors.
+// Licensed under the MIT License.
+
+#include <limits.h>
+#include <openenclave/host.h>
+#include <openenclave/internal/error.h>
+#include <openenclave/internal/tests.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include "../host/sgx/cpuid.h"
+#include "pf_gp_exceptions_u.h"
+
+static bool _is_misc_region_supported()
+{
+    uint32_t eax, ebx, ecx, edx;
+    eax = ebx = ecx = edx = 0;
+
+    // Obtain feature information using CPUID
+    oe_get_cpuid(CPUID_SGX_LEAF, 0x0, &eax, &ebx, &ecx, &edx);
+
+    // Check if EXINFO is supported by the processor
+    if (!(ebx & CPUID_SGX_MISC_EXINFO_BIT))
+        return false;
+    else
+        return true;
+}
+
+int main(int argc, const char* argv[])
+{
+    oe_result_t result;
+    oe_enclave_t* enclave = NULL;
+    int return_value;
+
+    if (argc != 2)
+    {
+        fprintf(stderr, "Usage: %s ENCLAVE_PATH\n", argv[0]);
+        return 1;
+    }
+
+    const uint32_t flags = oe_get_create_flags();
+
+    result = oe_create_pf_gp_exceptions_enclave(
+        argv[1], OE_ENCLAVE_TYPE_SGX, flags, NULL, 0, &enclave);
+    /* The enclave creation should succeed on both SGX1 and SGX2 machines. */
+    if (result != OE_OK)
```

Comment:
Should enclave creation succeed on SGX1 machines? If the enclave relies on PFGP handling, then it cannot possibly work on SGX1 machines. Therefore isn't it better to fail early at enclave creation itself?

Reply:
Depending on the usage, it might be acceptable for the SGX enclave to abort operation when the trust-worthy #PF/#GP info is not available inside the Enclave (for example, on SGX1) and to attempt to handle #PF/#GP inside the enclave when the trust-worth info is available. For enclaves that can't function without trust-worthy #PF/#GP info, and the developer would rather to fail the enclave load on SGX1 CPU, the developer can set SIGSTRUCT.MISCSELECT and SIGSTRUCT.MICSMASK to reflect the policy of failing Enclave initialization if #FP/#GP MISCSELECT is not enabled, either because the untrusted loader on SGX2 CPU does not follow developer request to enable #FP/#GP MISCSELECT, or the CPU does not support #FP/#GP MISCSELECT.

Reply:
@anakrish  @bodzhang I can see the two options that we can provide to developers
1. once opt-in, the enclave can run on both SGX1 and SGX2 CPU, but the #PF/#GP info will only be available on the latter
2. once opt-in, the enclave is only allowed to run on SGX2 CPU

The PR currently implement 1 only. Do you think we need to support both?

Reply:
Given, @bodzhang 's answer above, #1 is sufficient. The enclave author can get the behavior of #2 by using the SIGSTRUCT.MISCSELECT and SIGSTRUCT.MISCMASK fields.